src/attribution/onINP.ts

```diff
@@ -25,38 +28,224 @@ import {
+  });
+
+  // Remove all pending LoAF entries that don't intersect with entries in
+  // the newly cleaned up `pendingEntriesGroupMap`.
+  const loafsToKeep: Set<PerformanceLongAnimationFrameTiming> = new Set();
+  pendingEntriesGroupMap.forEach((group) => {
+    getIntersectingLoAFs(group.startTime, group.processingEnd).forEach(
+      (loaf) => {
+        loafsToKeep.add(loaf);
+      },
+    );
+  });
+  pendingLoAFs = Array.from(loafsToKeep);
+
+  // Reset the idle callback so it can be queued again.
+  idleCallback = -1;
+};
+
+const getIntersectingLoAFs = (
+  start: DOMHighResTimeStamp,
+  end: DOMHighResTimeStamp,
+) => {
+  const intersectingLoAFs = [];
+
+  for (let i = 0, loaf; (loaf = pendingLoAFs[i]); i++) {
+    // If the LoAF ends before the given start time, ignore it.
+    if (loaf.startTime + loaf.duration < start) continue;
+
+    // If the LoAF starts after the given end time, ignore it and all
+    // subsequent pending LoAFs (because they're in time order).
+    if (loaf.startTime > end) break;
+
+    // Still here? If so this LoAF intersects with the interaction.
+    intersectingLoAFs.push(loaf);
   }
-  // Set an empty object if no other attribution has been set.
-  (metric as INPMetricWithAttribution).attribution = {};
+  return intersectingLoAFs;
+};
+
+const attributeINP = (metric: INPMetric): void => {
+  const firstEntry = metric.entries[0];
+  const renderTime = entryToRenderTimeMap.get(firstEntry)!;
+  const group = pendingEntriesGroupMap.get(renderTime)!;
+
+  const processedEventEntries = group.entries;
+  const processingStart = firstEntry.processingStart;
+  const processingEnd = group.processingEnd;
+
+  const longAnimationFrameEntries: PerformanceLongAnimationFrameTiming[] =
+    getIntersectingLoAFs(firstEntry.startTime, processingEnd);
+
+  // The first interaction entry may not have a target defined, so use the
+  // first one found in the entry list.
+  // TODO: when the following bug is fixed just use `firstInteractionEntry`.
+  // https://bugs.chromium.org/p/chromium/issues/detail?id=1367329
+  const firstEntryWithTarget = metric.entries.find((entry) => entry.target);
+
+  // Since entry durations are rounded to the nearest 8ms, taking the average
+  // of a group of entries that we know were rendered in the same animation
+  // frame should give a better estimate of the "true" render time.
+  const AveragePresentationTime =
+    processedEventEntries.reduce((acc, entry) => {
+      return acc + (entry.startTime + entry.duration);
+    }, 0) / processedEventEntries.length;
+
+  const nextPaintTimeCandidates = [
+    AveragePresentationTime,
+    processingEnd,
+  ].concat(
+    longAnimationFrameEntries.map((loaf) => loaf.startTime + loaf.duration),
+  );
+
+  // To get the most accurate next paint time, use signals from the
+  // Event Timing API and LoAF to make the best possible guess.
+  const nextPaintTime = Math.max.apply(Math, nextPaintTimeCandidates);
```

Comment:
Out of curiosity why do you apply `Math` as the firts argument instead of `null`?

Reply:
You could do either, but since the `this` context in a normal call to `Math.max()` would be `Math`, it seemed to make sense to use that.

Reply:
Interestingly it looks like the [examples on MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/apply) use `null` (though they don't explain why). TBH I don't think it matters at all and presumably they both minify/compress to the same code size.

Reply:
Yeah that's why I asked after checking there! Both seem to work but was curious.

Reply:
Let's keep it as `Math` (I do think that's "more correct"). FWIW I'm 99% sure I remember a time when calling `console.log.apply(null, ['Hello World!])` would error because it didn't have the correct `this` context, so you had to do `console.log.apply(console, ['Hello World!])`.

Reply:
Nit: Is spread operator less supported? `Math.max(...nextPaintTimeCandidates)`

Reply:
I didn't use the spread operator here (or any other iterator stuff, e.g. for...of) because then when the bundle is transpiled to ES5 it'll include tons of Babel helper code bloat.

At some point in the future I think we can decide to not transpile to ES5, but I'd want to be intentional about that change and make sure we're not going to be breaking anyone.